adds a check for moveloops to check if the loop it tries to add already exists

MNarath1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This basically adds a proc that will check if the arguments forwarded to generate the new moveloop are identical with the ones on maybe an old loop before it allows it to overwrite it that way we won't endlessly make new loops and destroy old ones even trough there is no reason to.
Also fixes closes https://github.com/BeeStation/BeeStation-Hornet/issues/6844
turned out goliaths have a move delay of 4 Seconds enough time for the handle automatic movement proc to keep on overwriting the moveloop causing it to basically never allow the moveloop to fire
## Why It's Good For The Game
Bug fix good

## Changelog
:cl:
add: adds a check to ensure we won't keep on overwriting basically the same moveloop
fix: fixes mobs that would have an insane long move delay thus overwriting their moveloop before it could fire
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MNarath1 committed:
fix

MNarath1 committed:
instead we gonna check every time if this movement loop already exists

MNarath1 commented:
also fixes a comment mistake in the original PR as pointed out by what i assume was @LemonInTheDark on the tg discord

MNarath1 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -133,6 +140,18 @@
+		return FALSE
+
+
+
```
```suggestion
```

MNarath1 committed:
Update code/controllers/subsystem/movement/movement_types.dm

MNarath1 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -133,6 +140,15 @@
+	if(direction == dir)
+		return TRUE
+	else
+		return FALSE
```
```suggestion
		return FALSE

```

MNarath1 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -42,6 +42,13 @@
+	else
+		return FALSE
+
+
```
```suggestion

```

MNarath1 committed:
Update code/controllers/subsystem/movement/movement_types.dm

MNarath1 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -133,6 +138,15 @@
+	if(direction == dir)
+		return TRUE
+	else
+		return FALSE
```
```suggestion
		return FALSE

```

MNarath1 committed:
Update code/controllers/subsystem/movement/movement_types.dm

MNarath1 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -42,6 +42,11 @@
+	if(loop_type == type && priority == src.priority && flags == src.flags && delay == src.delay && timeout == lifetime)
+		return TRUE
+	else
+		return FALSE
```
```suggestion
		return FALSE

```

MNarath1 committed:
Update code/controllers/subsystem/movement/movement_types.dm

ike709 commented on code/controllers/subsystem/movement/move_handler.dm:
```diff
@@ -76,10 +76,13 @@ SUBSYSTEM_DEF(move_manager)
 /datum/movement_packet/proc/add_loop(datum/controller/subsystem/movement/subsystem, datum/move_loop/loop_type, priority, flags, datum/extra_info)
 	var/datum/move_loop/existing_loop = existing_loops[subsystem]
+
+	if(existing_loop && existing_loop.compare_vars(arglist(args.Copy(2))))
```
```suggestion
	if(existing_loop?.compare_vars(arglist(args.Copy(2))))
```

MNarath1 committed:
Update code/controllers/subsystem/movement/move_handler.dm

Co-authored-by: ike709 <ike709@users.noreply.github.com>

ike709 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -133,6 +139,16 @@
 
+/datum/move_loop/move/compare_vars(datum/move_loop/loop_type, priority, flags, extra_info, delay, timeout, dir)
+	. = ..()
+	if(!.)
```
```suggestion
	if(!..())
```

ike709 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -208,6 +224,16 @@
 
+/datum/move_loop/has_target/compare_vars(datum/move_loop/loop_type, priority, flags, extra_info, delay, timeout, atom/chasing)
+	. = ..()
+	if(!.)
```
```suggestion
	if(!..())
```

ike709 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -334,6 +360,16 @@
 
+/datum/move_loop/has_target/jps/compare_vars(datum/move_loop/loop_type, priority, flags, extra_info, delay, timeout, atom/chasing, repath_delay, max_path_length, minimum_distance, obj/item/card/id/id, simulated_only, turf/avoid, skip_first)
+	. = ..()
+	if(!.)
```
```suggestion
	if(!..())
```

ike709 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -505,6 +541,16 @@
 
+/datum/move_loop/has_target/dist_bound/compare_vars(datum/move_loop/loop_type, priority, flags, extra_info, delay, timeout, atom/chasing, dist = 0)
+	. = ..()
+	if(!.)
```
```suggestion
	if(!..())
```

ike709 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -657,6 +703,16 @@
 
+/datum/move_loop/has_target/move_towards/compare_vars(datum/move_loop/loop_type, priority, flags, extra_info, delay, timeout, atom/chasing, home = FALSE)
+	. = ..()
+	if(!.)
```
```suggestion
	if(!..())
```

ike709 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -797,6 +853,16 @@
 
+/datum/move_loop/move_rand/compare_vars(datum/move_loop/loop_type, priority, flags, extra_info, delay, timeout, list/directions)
+	. = ..()
+	if(!.)
```
```suggestion
	if(!..())
```

ike709 commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -42,6 +42,12 @@
 	return TRUE
 
+/datum/move_loop/proc/compare_vars(datum/move_loop/loop_type, priority, flags, extra_info, delay = 1, timeout = INFINITY) //exists so we can check if this exact moveloop datum already exists (in terms of vars) and so we can stop it from needlessly create a new one to overwrite
+	if(loop_type == type && priority == src.priority && flags == src.flags && delay == src.delay && timeout == lifetime)
```
```suggestion
/datum/move_loop/proc/compare_vars(datum/move_loop/loop_type, priority, flags, extra_info, delay = 1, timeout = INFINITY) //exists so we can check if this exact moveloop datum already exists (in terms of vars) and so we can stop it from needlessly create a new one to overwrite
	SHOULD_CALL_PARENT(TRUE)
	if(loop_type == type && priority == src.priority && flags == src.flags && delay == src.delay && timeout == lifetime)
```

MNarath1 committed:
Apply suggestions from code review

Co-authored-by: ike709 <ike709@users.noreply.github.com>

MNarath1 committed:
cleans up the code a bit now that i had sleep

MNarath1 commented:
with the last commit i should have reduced the ammount of if's in those procs a bit and condensed it into a single if that just uses basic && shortcut behavior

MNarath1 committed:
forgot that one

LemonInTheDark commented:
> also fixes a comment mistake in the original PR as pointed out by what i assume was @LemonInTheDark on the tg discord

The world may never know ahhhhh wooooo

LemonInTheDark commented on code/controllers/subsystem/movement/move_handler.dm:
```diff
@@ -76,10 +76,13 @@ SUBSYSTEM_DEF(move_manager)
+		return //it already exists stop trying to make the same moveloop
+
 	if(existing_loop && existing_loop.priority > priority)
 		if(!(existing_loop.flags & MOVEMENT_LOOP_IGNORE_PRIORITY) && !(flags & MOVEMENT_LOOP_IGNORE_PRIORITY))
```
Should do the priority check first, since it checks close to the same thing and is a bit less wild

MNarath1 replied:
While i would not say its the exact same but yeah  makes sense to do it first since the priority check doesn't need a whole new proc to work

LemonInTheDark commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -42,6 +42,11 @@
 	src.lifetime = timeout
 	return TRUE
 
+/datum/move_loop/proc/compare_loops(datum/move_loop/loop_type, priority, flags, extra_info, delay = 1, timeout = INFINITY) //exists so we can check if this exact moveloop datum already exists (in terms of vars) and so we can stop it from needlessly create a new one to overwrite
```
You should make this comment a doc comment (assuming bee has those)

MNarath1 replied:
bee doesn't trough i can see it beeing a bit clearer to see if i use it with that format (in contrast to same line comments)

LemonInTheDark commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -797,6 +822,10 @@
 
+/datum/move_loop/move_rand/compare_loops(datum/move_loop/loop_type, priority, flags, extra_info, delay, timeout, list/directions)
+	if(..() && potential_directions == directions) //i guess this could be usefull if actually it really has yet to move
+		return TRUE
```
A: usefull
B: This check will fail when you want it to succeed
```
var/list/a = list(1)
var/list/b = list(1)
// Even though these two lists have the same contents, they are different "objects"
world.log << (a == b) // So this will print 0, or FALSE
// What you really want to do is check if the two have the same contents
// Could do this by iterating, but why iterate when you can make byond do it
// Two different ways to do that

// This uses the "equality" operator https://www.byond.com/docs/ref/#/operator/~= 
//which does a shallow comparison between the two. 
// This may not actually be what you want tho, since list(1,2) is treated the same as list(2,1)
world.log << (a ~= b)

// BYOND allows for the use of binary arithmetic operators, like &, |, or ^ on lists
// | in this case "iterates" both lists, 
// and returns a new list that contains all unique items from both lists https://www.byond.com/docs/ref/#/operator/| 
world.log << (length(a | b)  == length(a))
```

MNarath1 replied:
right potential_directions is a list whops

MNarath1 replied:
> 

whoa this actually did not load for me first thats really useful

MNarath1 committed:
fix

some comment changes actually makes the rand comparison work and makes the priority check evaluate first

MNarath1 committed:
this is way better

i found the previous implementation really ugly so this is more than just a bit welcome

MNarath1 committed:
lets compare it to potential_directions instead

if we do it the other way around some issues could arise

MNarath1 committed:
adds a line
